File: Java/e_learn/src/main/java/gr/aueb/elearn/chapter17/collections/Product.java
package gr.aueb.elearn.chapter17.collections;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@ToString
@EqualsAndHashCode
public class Product implements Comparable<Product>{
    private String description;
    private Double price;
    private Integer quantity;
    
    @Override
    public int compareTo(Product o) {
        return description.compareTo(o.description);
    }

    // @Override
    // public int compareTo(Product o) {
    //     return this.getQuantity() - o.getQuantity();
    // }
    
    // @Override
    // public int compareTo(Product o) {
    //     return this.getQuantity().compareTo(o.getQuantity());
    // }

}

File: Csharp/SebAuebAug2022/ComparerApp/Product.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ComparerApp
{
    internal class Product : IComparable<Product>
    {
        public string? Description { get; set; }
        public double Price { get; set; }
        public int Quantity { get; set; }

        public Product() { }

        public Product(string? description, double price, int quantity)
        {
            Description = description;
            Price = price;
            Quantity = quantity;
        }

        public int CompareTo(Product? other)
            => this.Description!.CompareTo(other!.Description);

        public override string? ToString()
            => $"{this.Description} {this.Price} {this.Quantity}";

    }
}
